### PR TITLE
Use custom class for handling storage URIs

### LINKED ISSUE #8328: [Bug]: Nessie GC fails to handle Iceberg column names with quotes in S3
### What happened
I created using spark and iceberg a table X with columns `"_id"`,`B`, `C`
partitioned by truncate(2000, `"_id"`)
When inserting data into the table, under the data folder of the table in S3, folders of this kind were created:
s3://bucket/folder/X_12c98633-5492-48b4-950f-ec23497db481/data/"_id"_trunc=0/
After running the GC tool, I received an exception:
Caused by: java.lang.IllegalArgumentException: Illegal character in path at index 63.
at java.base/java.net.URI.create(URI.java:906)
at org.projectnessie.gc.iceberg.IcebergContentToFiles.dataFileUri(IcebergContentToFiles.java:262)
at org.projectnessie.gc.iceberg.IcebergContentToFiles.lambda$allDataFiles$3(IcebergContentToFiles.java:158)

### How to reproduce it
1)
CREATE TABLE X (
    `"_id"` bigint NOT NULL COMMENT 'unique id',
    A big int,
    B big int)
partitioned by truncate(2000, `"_id"`)
2)
insert into X values (100, 1, 1)
3)
drop table X
4) 
Run the gc tool with gc option to leave only the the last commit.

### Nessie server type (docker/uber-jar/built from source) and version

Spark 3.4
Iceberg 1.3.0
Nessie 0.70.0

### Client type (Ex: UI/Spark/pynessie ...) and version
Spark 3.4 (When creating the table and writing the data) with Iceberg tables.
Ran the GC tool from command line.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by dimas-b**
... because some S3 URIs are not parseable by java.net.URI

Fixes #8328

### CODE REVIEW

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    "file:opaque/a,file:path",
+    "file:opaque,file:/path",
+    "file:opaque/a,file:/path",
+    "s3://b/path,s3://b/path/file,file",
+    "s3://b/path/,s3://b/path/file,file",
+    "s3://b/path/,s3://b///path/file,file",
+    "s3://b/,s3://b/path/file,path/file",
+    "s3://b,s3://b/path/file,path/file",
+  })
+  void testRelativize(String base, String other) {
```
Comment (snazy) [inline tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:160]: The third parameter in the CSVs is unused??

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import java.nio.file.Paths;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class TestStorageUri {
+
+  private static URI normalizedURI(String value) {
+    URI u = URI.create(value).normalize();
```
Comment (snazy) [inline tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:33]: Why not just `return URI.create(value).normalize();` and omit the rest?

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    "s3://b/base/a,/absolute",
+    "s3://b/base/a/,/absolute",
+    "s3://b/base/a/,////absolute",
+    "s3://b/,file",
+    "s3://b,file/a",
+    "s3://b,file/a/",
+    "s3://b/,////absolute",
+    "s3://b////,////absolute",
+    "s3://b/base/,s3://b/file",
+  })
```
Comment (snazy) [inline tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:195]: Please add:
```
    "s3://b,/file/a/",
    "s3://b/foo,/file/a/",
    "s3://b/base/,s3://c/file",
```

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    "file:/ba,file:/base/file",
+    "file:opaque/a,path",
+    "file:opaque/a,file:path",
+    "file:opaque,file:/path",
+    "file:opaque/a,file:/path",
+    "s3://b/path,s3://b/path/file,file",
+    "s3://b/path/,s3://b/path/file,file",
+    "s3://b/path/,s3://b///path/file,file",
+    "s3://b/,s3://b/path/file,path/file",
+    "s3://b,s3://b/path/file,path/file",
```
Comment (snazy) [inline tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:158]: Please add:
```
    "s3://b,s3://c/path/file,path/file",
    "s3://b,file://c/path/file,path/file",
```

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        "file:///path/file/",
+        "file:///path/file",
+        "file://///path/file",
+        "file:///path/file//",
+        "file:///path////file//",
+        "s3:/path",
+        "s3://bucket",
+        "s3://bucket/",
+      })
+  void testLocation(String input) {
```
Comment (snazy) [inline tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:109]: Proposal for testing w/ a `StorageUri.path()`:
```java
  @ParameterizedTest
  @CsvSource({
        "/path/file,/path/file",
        "relative/file,relative/file",
        "file:/,/",
        "file:/path,/path",
        "file:/path/file,/path/file",
        "file:/path/file//,/path/file/",
        "file:///path/file/,/path/file/",
        "file:///path/file,/path/file",
        "file://///path/file,/path/file",
        "file:///path/file//,/path/file/",
        "file:///path////file//,/path/file/",
        "s3:/path,/path",
        "s3://bucket,",
        "s3://bucket/,/",
      })
  void testLocation(String input, String path) {
    StorageUri s = StorageUri.of(input);
    URI u = normalizedURI(input);
    assertThat(s).extracting(StorageUri::path, StorageUri::location, StorageUri::toString).containsExactly(path, u.toString(), u.toString());
  }
```

File: tools/storage/uri/README.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# Storage URI Utilities
+
+This module contains utilities for dealing with Object Storage URIs, which in many cases do not follow
+the [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) syntax.
+
+In particular, Amazon S3 URIs permit unescaped quote characters.
```
[the remark this answers is not in the document]
Comment (snazy) [reply tools/storage/uri/README.md:6]: oh

File: tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+public class StorageUri implements Comparable<StorageUri> {
+  public static final Comparator<StorageUri> COMPARATOR =
+      Comparator.<StorageUri, String>comparing(u -> normalizedForCompare(u.scheme))
+          .thenComparing(u -> normalizedForCompare(u.host))
+          .thenComparing(u -> normalizedForCompare(u.path));
+
+  private static final String SCHEME_FILE = "file";
+
+  private final String scheme;
+  private final String host;
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java:40]: good point :facepalm:

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    assertThat(StorageUri.of(input).toString()).isEqualTo(expected);
+  }
+
+  @Test
+  void testLocalPath() {
+    Path local = Paths.get("/path/file");
+    assertThat(StorageUri.of("file:/path/file").toLocalPath()).isEqualTo(local);
+    assertThat(StorageUri.of("file:///path/file").toLocalPath()).isEqualTo(local);
+
+    assertThatThrownBy(() -> StorageUri.of("/unspecified").toLocalPath())
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:132]: `IcebergContentToFiles.checkUri(...)` takes care of that, doesn't it?

File: tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+    Preconditions.checkArgument(
+        !path.startsWith("."), "Parent and self-references are not supported: %s", path);
+
+    String base = normalizedBase(location());
+    path = normalizedPath(path);
+
+    if (path.startsWith("/")) { // absolute path
+      int pos = base.indexOf("://");
+      if (pos >= 0) { // `base` has a host element
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java]: this code has been reworked. Please re-review :)

File: gc/gc-base/src/main/java/org/projectnessie/gc/expire/PerContentDeleteExpired.java [hunk 10]
```diff
@@ -216,14 +217,8 @@ BloomFilter<URI> createBloomFilter() {
...
-  private static void funnel(URI uri, PrimitiveSink sink) {
-    funnelString(uri.getScheme(), sink);
-    funnelString(uri.getHost(), sink);
-    int port = uri.getPort();
-    if (port != 0) {
-      sink.putInt(uri.getPort());
-    }
-    funnelString(uri.getRawPath(), sink);
+  private static void funnel(StorageUri uri, PrimitiveSink sink) {
+    funnelString(uri.location(), sink);
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-base/src/main/java/org/projectnessie/gc/expire/PerContentDeleteExpired.java:221]: Changed the java doc and method body here as well.

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java (hunk 2 above)
Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:33]: The `URI` class remembers the original string and only makes normalized components available via getters.

Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:33]: added a comment

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java (hunk 5 above)
Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:109]: Added a separate method for this

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java (hunk 4 above)
Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:158]: still needed?

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java (hunk 1 above)
Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java:160]: fixed

File: gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/IcebergContentToFiles.java [hunk 11]
```diff
@@ -191,91 +194,88 @@ static Stream<URI> allDataAndDeleteFiles(
    * All processed {@link URI}s must have a schema part and, if the schema is {@code file}, the path
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/IcebergContentToFiles.java]: fixed

File: tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    // drop trailing path separators
+    while (loc.endsWith("/")) {
+      loc = loc.substring(0, loc.length() - 1);
+    }
+    return loc;
+  }
+
+  private static String normalizedPath(String loc) {
+    // drop excessive leading path separators
+    while (loc.startsWith("//")) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java]: Added test cases.

File: tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+  @Nullable
+  public String scheme() {
+    return scheme;
+  }
+
+  public boolean isLocalFile() {
+    return SCHEME_FILE.equalsIgnoreCase(scheme());
+  }
+
+  public Path toLocalPath() {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java:134]: moving there

[the remark this answers is not in the document]
Comment (snazy) [reply tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java:134]: don't see it there?

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (snazy) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java]: It does. It's just handled in StorageUri now

File: gc/gc-base/src/main/java/org/projectnessie/gc/files/local/LocalFiles.java [hunk 14]
```diff
@@ -84,10 +83,10 @@ public DeleteResult delete(FileReference fileReference) {
     }
   }
 
-  static URI ensureTrailingSlash(URI uri) {
-    if (uri.getPath().endsWith("/")) {
+  static StorageUri ensureTrailingSlash(StorageUri uri) {
+    if (uri.location().endsWith("/")) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-base/src/main/java/org/projectnessie/gc/files/local/LocalFiles.java:87]: refactoring

File: tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+    Preconditions.checkArgument(isLocalFile(), "Not a local file: %s", location());
+    return Paths.get(URI.create(location()));
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java:136]: leftover from an early version :facepalm: refactoring into `LocalFiles`

File: gc/gc-iceberg-files/src/main/java/org/projectnessie/gc/iceberg/files/IcebergFiles.java [hunk 16]
```diff
@@ -63,14 +63,14 @@ public static Builder builder() {
...
 
   static Stream<String> filesAsStrings(Stream<FileReference> fileObjects) {
-    return fileObjects.map(FileReference::absolutePath).map(URI::toString);
+    return fileObjects.map(FileReference::absolutePath).map(StorageUri::location);
   }
 
-  static URI ensureTrailingSlash(URI uri) {
-    if (uri.getPath().endsWith("/")) {
+  static StorageUri ensureTrailingSlash(StorageUri uri) {
+    if (uri.location().endsWith("/")) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-iceberg-files/src/main/java/org/projectnessie/gc/iceberg/files/IcebergFiles.java:70]: reworked

File: gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/ITSparkIcebergNessieLocalNoSchema.java [hunk 17]
```diff
@@ -108,17 +108,18 @@ public void roundTripLocal() throws Exception {
...
       soft.assertThat(deleteSummary.deleted()).isEqualTo(3L);
       filesAfter = allFiles(icebergFiles);
-      Set<URI> removedFiles = new HashSet<>(filesBefore);
+      Set<StorageUri> removedFiles = new HashSet<>(filesBefore);
       removedFiles.removeAll(filesAfter);
       // The first and second table-metadata and the manifest-list from the second table metadata
       // got cleaned up.
       soft.assertThat(removedFiles)
-          .allMatch(u -> u.getPath().endsWith(".json") || u.getPath().endsWith(".avro"));
+          .allMatch(u -> u.location().endsWith(".json") || u.location().endsWith(".avro"));
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/ITSparkIcebergNessieLocalNoSchema.java:116]: It's a test... `path()` is nullable, it leads to more code :shrug:

File: tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java]: `toLocalPath()` has been moved to `LocalFiles`

File: gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java [hunk 18]
```diff
@@ -147,6 +147,17 @@ static Stream<TestCase> testCases() {
             .addSteps(dml("INSERT INTO nessie.tc_2.tbl_a select 24, \"case\""))
             .putPolicies("main", numCommits(1))
             .putPolicies("branch_tc_2", numCommits(1))
+            .build(),
+        // 3
+        testCase()
+            .namespace("tc_3")
+            .addSteps(
+                expiredDdl(
+                    "CREATE TABLE nessie.tc_3.tbl_a (`id\"` int, name string) "
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java:156]: will do presently

File: gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/IcebergContentToFiles.java [hunk 19]
```diff
@@ -207,7 +209,7 @@ static StorageUri checkUri(String type, String location, ContentReference conten
       return StorageUri.of("file://" + location);
     }
 
-    if (loc.isLocalFile()) {
+    if (SCHEME_FILE.equals(loc.scheme())) {
```
Comment (snazy) [inline gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/IcebergContentToFiles.java:212]: Nit (non-blocking): I think, we can remove that constant in StorageUri. `"file"` here would be good enough

File: gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/ITSparkIcebergNessieLocalNoSchema.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (snazy) [reply gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/ITSparkIcebergNessieLocalNoSchema.java:116]: for a test - fair enough

File: gc/gc-base/src/test/java/org/projectnessie/gc/roundtrip/TestMarkAndSweep.java [hunk 20]
```diff
@@ -286,7 +286,7 @@ public void close() {}
         };
     FileDeleter deleter =
         fileObject -> {
-          soft.assertThat(fileObject.path().getPath()).endsWith("-unused");
+          soft.assertThat(fileObject.path().location()).endsWith("-unused");
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-base/src/test/java/org/projectnessie/gc/roundtrip/TestMarkAndSweep.java:289]: as discussed in a similar comment, `path()` is nullable - too much boiler plate for test code

File: gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (dimas-b) [reply gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java]: I wanted to use special charts in partition values, but they seem to work fine :)

So I added a bunch of other special chars to this test.

Somehow `#` still breaks it... but I can probably debug this later.

File: gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java [hunk 21]
```diff
@@ -153,8 +153,10 @@ static Stream<TestCase> testCases() {
             .namespace("tc_3")
             .addSteps(
                 expiredDdl(
-                    "CREATE TABLE nessie.tc_3.tbl_a (`id\"` int, name string) "
-                        + "PARTITIONED BY (`id\"`)"))
+                    // Note: intentional special chars in the column name
+                    // TODO: debug why `#` breaks this test
```
Comment (snazy) [inline gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java:157]: URI fragment?

Comment (snazy) [reply gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java:157]: Also, let's have special characters in table and namespace names.
I'm okay to have those in a follow-up.

**Review by snazy — APPROVED** [review]

File: gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java (hunk 21 above)
Comment (dimas-b) [reply gc/gc-iceberg-inttest/src/intTest/java/org/projectnessie/gc/iceberg/inttest/AbstractITSparkIcebergNessieObjectStorage.java:157]: Apparently... but I'd like to check whether AWS chokes on it or MinIO or Iceberg... will probably have to do that tomorrow.

**Pull request merged**